src/coreclr/jit/hwintrinsicarm64.cpp

```diff
@@ -1786,6 +1787,65 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
             break;
         }
 
+        case NI_AdvSimd_StoreSelectedScalar:
+        case NI_AdvSimd_Arm64_StoreSelectedScalar:
+        {
+            assert(sig->numArgs == 3);
+            assert(retType == TYP_VOID);
+
+            CORINFO_ARG_LIST_HANDLE arg1     = sig->args;
+            CORINFO_ARG_LIST_HANDLE arg2     = info.compCompHnd->getArgNext(arg1);
+            CORINFO_ARG_LIST_HANDLE arg3     = info.compCompHnd->getArgNext(arg2);
+            var_types               argType  = TYP_UNKNOWN;
+            CORINFO_CLASS_HANDLE    argClass = NO_CLASS_HANDLE;
+            argType                = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg3, &argClass)));
+            op3                    = impPopStack().val;
+            argType                = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg2, &argClass)));
+            op2                    = impPopStack().val;
+            unsigned fieldCount    = info.compCompHnd->getClassNumInstanceFields(argClass);
+            int      immLowerBound = 0;
+            int      immUpperBound = 0;
+
+            if (op2->TypeGet() == TYP_STRUCT)
+            {
+                info.compNeedsConsecutiveRegisters = true;
+
+                if (!op2->OperIs(GT_LCL_VAR))
+                {
+                    unsigned tmp = lvaGrabTemp(true DEBUGARG("StoreSelectedScalarN"));
+
+                    impStoreTemp(tmp, op2, CHECK_SPILL_NONE);
+                    op2 = gtNewLclvNode(tmp, argType);
+                }
+                op2 = gtConvertTableOpToFieldList(op2, fieldCount);
+            }
+            else
+            {
+                // While storing from a single vector, both Vector128 and Vector64 API calls are in AdvSimd class.
+                // Thus, we get simdSize as 8 for both of the calls. We re-calculate that simd size for such API calls.
+                getBaseJitTypeAndSizeOfSIMDType(argClass, &simdSize);
+            }
+
+            assert(HWIntrinsicInfo::isImmOp(intrinsic, op3));
+            HWIntrinsicInfo::lookupImmBounds(intrinsic, simdSize, simdBaseType, &immLowerBound, &immUpperBound);
+            op3     = addRangeCheckForHWIntrinsic(op3, immLowerBound, immUpperBound);
```

Comment:
Should use `addRangeCheckIfNeeded` so we can also get some extra asserts checks present in it.

Reply:
cool... When I had added `addRangeCheckIfNeeded`, the native image generation step was failing on the following assert. May need some tweaks to the assert. I'll check this.
```
Assertion failed '!immOp->IsCnsIntOrI()' in 'System.Text.Unicode.Utf8Utility:TranscodeToUtf8(ulong,int,ulong,int,byref,byref):int' during 'Importation' (IL size 2297; hash 0x85587dde; FullOpts)
```

Reply:
Looks like `StoreSelectedScalar` for single struct needs to be handled properly. https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Transcoding.cs,1001

Reply:
you're right.. needed to move the range check after recomputing the `simdSize` 🙂. 
Is there anything we need to do to push this PR further?

Reply:
Seems https://github.com/dotnet/runtime/pull/93223/commits/142bcc08b2f9c23d558c6ea0ab63383ed66d5238 introduced these failures on Mono because it touches the existing `StoreSelectedScalar` API. AFAIK, they worked fine before your commit. I do not know much on how to debug/repro this, but can you double check?

```
/tmp/helix/working/B11909CC/p/build/apple/AppleBuild.targets(242,5): error : Precompiling failed for /tmp/helix/working/B11909CC/w/ADA40973/e/publish/System.Private.CoreLib.dll with exit code 139. [/private/tmp/helix/working/B11909CC/w/ADA40973/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B11909CC/p/build/apple/AppleBuild.targets(242,5): error :  [/private/tmp/helix/working/B11909CC/w/ADA40973/e/publish/ProxyProjectForAOTOnHelix.proj]
```